### PR TITLE
List Index fix, validation split in fit() and fit_chunked() for large datasets

### PR DESCRIPTION
**Pull request by aleju**
Hi,
I made a few changes to your code, maybe you want to reuse them.
They are:
1) Fixed an error "List indices must be integers, not list." that appeared whenever a list was called with a list of integers as the index. Google indicated that only Scipy seems to support that...? It required quite some changes to model.py. My tests indicated that everything still works.
2) I added a validation_split option to fit() in models.py to easily split of a part of the dataset for validation (e. g. validation_loss=0.2 to split of 20%). At the end of each epoch it calculates the loss on the validation set and prints it next to the loss on the training set, e. g.:
    Epoch 0
    2500000/2500000 [==============================] - 612s - loss: 0.1012 - val. loss: 0.1213
    Epoch 1
    2500000/2500000 [==============================] - 602s - loss: 0.0549 - val. loss: 0.0702
3) I added fit_chunked() to models.py which is designed to perform more easily a fit() on datasets that are too big for the GPU RAM or main memory. Basically instead of a dataset you provide a function that can be called to create a new generator for the dataset. That function is called at the start of each epoch to create a new generator and then the data is loaded from the generator in chunks that are small enough for the RAM. Then essentially the normal fit() is run on the chunk. That is repeated until all chunks from the dataset have been processed. Then the next epoch starts and the generator is "reset" by creating a new one.

### CODE REVIEW

**Comment by fchollet** [comment]
I'm going to integrate the validation split idea, it seems convenient to have as an option, rather than leaving it be to handled manually by the user. Cool stuff! : )

I'm not sure that fit_chunked is much needed: it is very easy to train/test a Keras model with a data generator:
```python
for X_batch, y_batch in generator:
   loss = model.train(X_batch, y_batch)
```
The cifar10_cnn.py example provides more detail on this. 

What's your justification for fit_chunked? I don't think including it would be in the spirit of keeping the codebase as simple and concise as possible. People who need this functionality can take care of it outside the model class in a couple lines of code.

**Comment by aleju** [comment]
I just like to have such functionality (or an example) in the library so that I can focus on the data and not write too much boiler plate code. It also shows clearly, that the library actually supports generators somehow and doesn't do funky stuff when you call the model's training method multiple times (e.g. reset all parameters at every call). I didn't see that there was already an example on how to do that with keras for the cifar dataset, so I guess my method doesn't have to be added.

**Comment by fchollet** [comment]
I integrated your validate_split contribution under your name (after some quick edits). https://github.com/fchollet/keras/commit/c57d5cce7903511edd4048f8bfed2ad0dc6f6b6b

**Comment by aptr322** [comment]
List index fix didn't make it to tree!

diff --git a/keras/models.py b/keras/models.py
index f4387ce..0d34f45 100644
--- a/keras/models.py
+++ b/keras/models.py
@@ -6,6 +6,7 @@ import optimizers
 import objectives
 import time, copy
 from utils.generic_utils import Progbar
+from math import ceil
 
 def standardize_y(y):
     y = np.asarray(y)
@@ -80,7 +81,7 @@ class Sequential(object):
             if shuffle:
                 np.random.shuffle(index_array)
 
-            nb_batch = len(X)/batch_size+1
+            nb_batch = int(ceil(float(len(X))/float(batch_size)))
             progbar = Progbar(target=len(X))
             for batch_index in range(0, nb_batch):
                 batch_start = batch_index*batch_size
